### PR TITLE
Fix failure to parse token errors as Cloud Errors - ARO-1577

### PR DESCRIPTION
**Pull request by s-fairchild**
### Which issue this PR addresses:

Fixes https://issues.redhat.com/browse/ARO-1577

### What this PR does / why we need it:

This PR updates all cases when `ParseUnverified` is called and returns a cloud error, which will then be converted to a `UserError`. This prevents the generic `InternalServerError` status code from being the only error.

### Test plan for issue:
Updated existing unit tests.

### Is there any documentation that needs to be updated for this PR?
I don't believe so.

### CODE REVIEW

File: pkg/util/instancemetadata/prod.go [hunk 1]
```diff
@@ -71,6 +72,7 @@ func (p *prod) populateTenantIDFromMSI(ctx context.Context) error {
 	c := &azureclaim.AzureClaim{}
 	_, _, err = parser.ParseUnverified(token.OAuthToken(), c)
 	if err != nil {
+		err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
 		return err
```
Comment (cadenmarchese) [inline pkg/util/instancemetadata/prod.go:76]:
```suggestion
		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
```

File: pkg/util/clusterauthorizer/authorizer.go [hunk 2]
```diff
@@ -71,6 +73,7 @@ func (a *azRefreshableAuthorizer) NewRefreshableAuthorizerToken(ctx context.Cont
 	c := &azureclaim.AzureClaim{}
 	_, _, err = p.ParseUnverified(token.OAuthToken(), c)
 	if err != nil {
+		err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
```
Comment (cadenmarchese) [inline pkg/util/clusterauthorizer/authorizer.go:76]: How did you confirm that all `token contains` errors come from `p.ParseUnverified(token.OAuthToken(), c)`? Mostly just curious.

File: pkg/api/validate/dynamic/serviceprincipal.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ func (dv *dynamic) ValidateServicePrincipal(ctx context.Context, clientID, clien
 	c := &azureclaim.AzureClaim{}
 	_, _, err = p.ParseUnverified(token.OAuthToken(), c)
 	if err != nil {
+		err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
 		return err
```
Comment (cadenmarchese) [inline pkg/api/validate/dynamic/serviceprincipal.go:29]:
```suggestion
		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
```

File: pkg/util/clusterauthorizer/authorizer.go (hunk 2 above)
Comment (s-fairchild) [reply pkg/util/clusterauthorizer/authorizer.go:76]: Searching for that string in our repo (excluding vendor/*) results in one result from a test. If you search for this string '"The provided service principal generated an invalid token."' you find one result in `aad.go`. The real failure to parse token error comes from [parser.go](https://github.com/Azure/ARO-RP/blob/master/vendor/github.com/form3tech-oss/jwt-go/parser.go#L99).

That makes me believe it's the source of all those errors.

File: pkg/util/instancemetadata/prod_test.go [hunk 4]
```diff
@@ -205,7 +205,7 @@ func TestPopulateTenantIDFromMSI(t *testing.T) {
 					OAuthToken().
 					Return("invalid")
 			},
-			wantErr: "token contains an invalid number of segments",
+			wantErr: "400: InvalidServicePrincipalToken: properties.servicePrincipalProfile: token contains an invalid number of segments",
```
Comment (rogbas) [inline pkg/util/instancemetadata/prod_test.go:208]: Can you please replace in the error message `token contains an invalid number of segments` with something that's more meaningful to end customers in this context, like `The provided service principal is invalid` ?

File: pkg/util/instancemetadata/prod.go [hunk 5]
```diff
@@ -71,7 +72,7 @@ func (p *prod) populateTenantIDFromMSI(ctx context.Context) error {
 	c := &azureclaim.AzureClaim{}
 	_, _, err = parser.ParseUnverified(token.OAuthToken(), c)
 	if err != nil {
-		return err
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", err.Error())
```
Comment (darthhexx) [inline pkg/util/instancemetadata/prod.go:75]: This function is called by the RP / Gateway /Monitor, etc. when they are first starting up and performing MSI authentication so there is no concept of an API or end user. This should stay as a regular Internal Server Error.

File: pkg/util/instancemetadata/prod_test.go (hunk 4 above)
Comment (s-fairchild) [reply pkg/util/instancemetadata/prod_test.go:208]: @rogbas Yes of course, I'll update the PR.

File: pkg/util/instancemetadata/prod.go (hunk 5 above)
Comment (s-fairchild) [reply pkg/util/instancemetadata/prod.go]: I changed it back to returning a regular error.

**Review by cadenmarchese — APPROVED** [review]
LGTM

**Pull request merged**